### DISCUSSION TITLE
Root drectory vs something like src directory

### DISCUSSION THREAD

**Discussion by luisfmelo** [benbjohnson/wtf#7] [General]
I want to know your opinion on using the root directory to place `event.go`, `auth.go`, etc. instead of using something like a `src`, `wtf`, `internal` or another naming.

I totally understand and I do it in the same way but, I also think that the root directory usually feels bloated because we have a lot of folders, a lot of `*.go` files, a lot of `*_test.go` files and we still need to put the Dockerfile, `go.mod`, some Jenkinsfile/travis, etc and other configuration files for coverage, ci/cd, etc.

**Comment by benbjohnson** [comment]
Hi Luís, good question.

I'm not a big fan of `src` as it's not too common in the Go community. I don't typically like `internal` either because I like to design the API for external packages from the beginning and moving `internal` code over to an exported package later on feels clunky.

Having a `wtf` subpackage is ok if you dislike a bloated root folder. I don't love it because it's redundant to type `wtf/wtf` for your package import but that's not terrible.

**Comment by luisfmelo** [comment]
Yeah, I agree. That's my dilemma: root package, be bloated or not be bloated.
By the way, nice work. If you find the time, it would be nice to have some blog posts to explain your thought process in creating this solution.

**Comment by lebzm** [comment]
Yes, finally someone thought on that matters too, I also have this dilemma and ended up with `wtf/wtf`, some redundancy in the `import` statement but the `import` statement is less likely to be read by Go developer (they read code below it, the real problem). For me, the most important thing is it still maintain the package name `wtf.User` and neat file structure (doesn't mix `.go` file with others).

**Comment by hexxfy** [comment]
I think `wtf/wtf` subfolder is a good option.
Even Rob Pike did the same in his own project called [Upspin](https://github.com/upspin/upspin):
`upspin.io` is the module name.
`upspin.io/upspin` is the subpackage for domain types.
Another option is to declare domain types directly in the internal package. That's how the Go team did in [pkgsite](https://github.com/golang/pkgsite) (the source code for the [pkg.go.dev](https://pkg.go.dev/) website).

For my server projects, I would place domain types either in the `internal` package directly or in the `internal/{projectname}` package, because server projects typically are not dedicated to exporting functionality. This was also mentioned on the official [go.dev](https://go.dev/doc/modules/layout#server-project) website. It also helps avoid a bloated root package.
